go.mod

```diff
@@ -343,7 +344,9 @@ replace (
 	// lock wasmvm so we do not break the grandpa contract
 	github.com/CosmWasm/wasmvm => github.com/CosmWasm/wasmvm v1.2.1
 
-	github.com/cosmos/ibc-apps/middleware/packet-forward-middleware/v7 => github.com/notional-labs/ibc-apps/middleware/packet-forward-middleware/v7 v7.0.0-20231027045618-f659894b12d9
+	// TODO: Move to notional once https://github.com/notional-labs/ibc-apps/pull/1 is merged.
```

Comment:
what's the issue with the upstream version of pfm? Why do we still need a fork?
Why can't we just migrate to https://github.com/notional-labs/ibc-apps/blob/main/middleware/packet-forward-middleware/go.mod

Reply:
I think the answer is dot But let me keep checking

Reply:
There was a code diff between the upstream version of pfm and since I did not fully understood the changes on the fork, I upgraded the fork but maintained the changes from notional 

The code diff was mostly related to parachain token info

https://github.com/RafilxTenfen/ibc-apps/blob/6c70c9f1396d2b90fcb7eda5de1014828ee09070/middleware/packet-forward-middleware/packetforward/keeper/keeper.go#L246

https://github.com/RafilxTenfen/ibc-apps/blob/6c70c9f1396d2b90fcb7eda5de1014828ee09070/middleware/packet-forward-middleware/packetforward/keeper/keeper.go#L313

Reply:
@RafilxTenfen I did the same thing

Reply:
I @faddat is ok with changes I am ok too. 

Afaik logic is next:

when token `transfer/channel-0/1` arrives from PICASSO parachain channel than its denom does not becomes ibc/HASH (see relevant ADRs), but it is translated to PICA. PICA minted.

when token PICA is sent to `transfer/channel-0` back on Picasso, reverse rewrite happens.

sending and receiving to all other chains works as usual. 
sending and receiving all other tokens works as usual.

things are easy here.

so what happens with PFM?

Reply:
When packet is received, it may have PFM or not. If it contains PFM, then receiver middle ware should rewrite IBC denom in case sender is transfer/channel-0.

IBC host calls TRANSFER middleware (rewrite denom) call PFM module, it sees new denom, it sends new denom via ICS04 adaper to other chain.

What is ICS04 addapter? It actually TRANSFER middleware. Because if PFM routes transfer back to Picasso, it should rewrite denom back. Atter that TRANSFER calls ICS04 adapter of ICS20.

So actually  TRANSFER middleware sandwiches PFM, not fork needed.

Reply:
@dzmitry-lahoda 

> when token transfer/channel-0/1 arrives from PICASSO parachain channel than its denom does not becomes ibc/HASH (see relevant ADRs), but it is translated to PICA. PICA minted.

Could you point me to the ADRs?

Reply:
i guess you know these

https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md
(and relevant spec links)

https://ibc.cosmos.network/main/ibc/middleware/integration

https://github.com/cosmos/ibc/tree/main/spec/app/ics-030-middleware

and relevant go code.